docs/_posts/2022-05-06-writing-cli-tools-with-typescript.md

```diff
@@ -0,0 +1,187 @@
+
+Preferably, we also want to have a single, packaged binary as the result of building our project rather than a script file to reduce installation overhead on part of the user. 
+
+To assure high code quality we also typically want to include code formatting, linting, and tests.
+
+## Basic development flow
+
+To hit the ground running, I created a starter project which already contains all required boilerplate and dependencies for running, formatting, linting, testing the code and packaging it to a binary. 
+
+> ℹ️ The starter project can be found on my [GitHub](https://github.com/fkurz/typescript-cliutil-starter) and via [npm](https://www.npmjs.com/package/@fkurz/typescript-cliutil-starter/). 
+
+```bash
+git clone git@github.com:fkurz/typescript-cliutil-starter.git my-cli
+cd my-cli 
+npm i
+```
+
+Our starter project uses 
+
+- the [*commander*](https://www.npmjs.com/package/commander) package for argument parsing and executing commands, 
+- code linting and formatting with [eslint](https://eslint.org/) and [prettier](https://prettier.io/), 
+- [Jest](https://jestjs.io/) as test runner and assertion library, and 
+- [pkg](https://github.com/vercel/pkg) to build a stand-alone binary.
+
+To illustrate the development flow, we now add an example subcommand and test it. 
+
+For example, we could add a command *yell* which is exactly like the *say* command (that already exists in the project) but a bit less subtle.
+
+To do this, we add a file *src/yell.ts* and then add the following code.
+
+```typescript
+// src/yell.ts
+import { Command } from "commander";
+
+export default (): Command =>
+ new Command()
+  .name("yell")
+  .description("Say the word passed as the first argument but louder")
+  .argument("<word>")
+  .action((word: string) => console.log(word.toLocaleUpperCase()));
+```
+
+
+
+We also have to register our yell command with the main CLI command in src/cmd.ts, i.e. add an import statement for our subcommand and register it using *Commander.addCommand* 
+
+```typescript
+// src/cmd.ts
+// … 
+import buildYellCmd from "./yell";
+// …
+export default (): Command => {
+ const command = new Command()
+  .option("-g, --greet", `Say ${HELLO}`, false)
+  .addCommand(buildSayCmd())
+  .addCommand(buildYellCmd())
+  .addHelpCommand()
+  .showHelpAfterError();
+```
+
+
+
+```
+```
+
+Having done that, we can run a quick test on our command using the dev script in package.json. The dev script simply executes our TypeScript code with [ts-node](https://github.com/TypeStrong/ts-node).
+
+```bash
+$ npm run dev -- yell 'hey!'
+
+> @fkurz/typescript-cliutil-starter@1.0.0 dev /private/tmp/my-cli
+> ts-node src/main.ts "yell" "hey!"
+
+HEY!
+```
+
+We can also add a quick test by adding *src/yell.test.ts* with the following content
+
+```typescript
+import buildCmd from "./cmd";
+
+beforeEach(() => {
+ jest.spyOn(process, "exit").mockImplementation();
+ jest.spyOn(console, "log").mockImplementation();
```

Comment:
`jest.spyOn()` already creates a mock function. There should be no need for an empty `.mockImplementation()` if you don't need to add a specific mock behavior.

```suggestion
 jest.spyOn(process, "exit");
 jest.spyOn(console, "log");
```

Reply:
I got his from SO. You're right though, the test also runs without `.mockImplementation` .